### PR TITLE
[JENKINS-67456] add check that pattern is valid

### PR DESCRIPTION
**Pull request by mawinter69**
performs a check of the pattern when the field looses focus
Additionally check when pressing the add button if the pattern is valid to avoid that an invalid pattern can be added at all.
see [JENKINS-67456](https://issues.jenkins.io/browse/JENKINS-67456)

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 1]
```diff
@@ -940,6 +941,18 @@ public boolean showPermission(String type, Permission p) {
         }
     }
 
+    @RequirePOST
+    @Restricted(NoExternalUse.class)
+    public FormValidation doCheckPattern(@QueryParameter String value) {
+      try {
+          @SuppressWarnings("unused")
+          Pattern pattern = Pattern.compile(value);
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java:949]:
```suggestion
          Pattern.compile(value);
```

File: src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/manage-project-roles.jelly [hunk 2]
```diff
@@ -148,6 +148,13 @@
             alert("Please enter a pattern");
             return;
           }
+          try {
+            new RegExp(pattern);
```
Comment (AbhyudayaSharma) [inline src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/manage-project-roles.jelly:152]: This is a really interesting idea. But there are differences between `java.util.regex` and JavaScript `RegExp`. See https://en.wikipedia.org/wiki/Comparison_of_regular_expression_engines

Comment (mawinter69) [reply src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/manage-project-roles.jelly]: ack

**Review by AbhyudayaSharma — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by kuoruan** [comment]
The `escape` function is **deprecated** https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape
Can we use [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) instead?

**Comment by NotMyFault** [comment]
> The `escape` function is **deprecated** https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape Can we use [encodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent) instead?

Interesting, feel free to submit a new PR addressing the deprecation. I'm not a maintainer of this plugin.

**Pull request merged**